### DISCUSSION TITLE
[API Request] getLyrics improvements

### DISCUSSION THREAD

**Discussion by kgarner7** [opensubsonic/open-subsonic-api#60] [API requests]
### Type of change

API extension

### Proposal description

This extension seeks to improve the usability of `getLyrics` in a few ways:
1. Add optional `id` parameter to just filter by the file ID. This gives precise control over the track to fetch (useful if you have same artist/title). Similarly, add an `artistId` field for querying artist by id.
2. Support explicitly requesting/returning synced/unsynced lyrics, similar to [Jellyfin](https://github.com/jellyfin/jellyfin/pull/8381). Additionally, support multiple languages/lyrics (and filter). This helps clients fetch synced/unsynced lyrics, and additionally enables XML API users to have proper newline support.
3. (Maybe?) Include the track id and artist id in the lyric response. This can be useful if you are just searching for lyrics by artist/title and want to be able to differentiate them

### Backward compatibility impact

~~Both options are optional, so the default if not provided **must** be the same. Potentially backwards-incompatible change would be adding `artistId`/`id` field to the resulting lyrics~~
None: this will be a new endpoint `/getLyrics2` (?)

### Backward compatibility

- [X] No backward compatibility impact.

### ID parameter
Add one **required string** parameters:
- `id`: the track id
- ~~`artistId`: the artist id~~ (decided against)

### Lyric output improvement
Adds two new parameters:
- ~~`preferSynced`: optional boolean parameter, denotes whether the server should prefer synced or not~~ (decided against: unnecessary since timestamps can simply be ignored for unsynced display)
- `lang`: optional string (repeatable), the preferred language of the lyrics (e.g., file has multiple lyrics). If included multiple times, search for lyrics matching **any** format.

The output **must** support multiple lyrics. For XML, this can be accomplished with multiple `lyrics` objects, for JSON `lyrics` should be an array. Every lyric **must** include the language.

~~If searching by `artist`/`artistId`/`title`, must return **all** matches.~~ (searchLyrics functionality rejected for now)

The server **must** return synchronous lyrics, if they exist (using Jellyfin format). ~~This is similar to `preferSynced` being false, with the addition that~~ The `line` **must** have a `start` field, denoting when to start displaying the given line measured as milliseconds from the start of the song (0 being right at the beginning). If lyrics exist, but they are unsynced, the server **must** return them, must have `synced` being false and **must** omit `start`.

###### Synced lyrics response

```xml
<subsonic-response status="ok" version="1.2.0" openSubsonic="true">
  <lyrics artist="Muse" title="Hysteria" synced="true" lang="en">
    <line start="0">It's bugging me</line>
    <line start="2000">Grating me</line>
    <line start="3001">And twisting me around...</line>
  </lyrics>
</subsonic-response>
```
```json
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.2.0",
    "openSubsonic": true,
    "lyrics": [{
      "artist": "Muse",
      "title": "Hysteria",
      "synced": true,
      "lang": "en",
      "line": [
        { "start": 0, "value": "It's bugging me"},
        { "start": 2000, "value": "Grating me" },
        { "start": 3001, "value": "And twisting me around..." }
      ]
    }]
  }
}
```

##### Unsynced lyrics response
```xml
<subsonic-response status="ok" version="1.2.0" openSubsonic="true">
  <lyrics artist="Muse" title="Hysteria" synced="false" lang="eng">
    <line>It's bugging me</line>
    <line>Grating me</line>
    <line>And twisting me around...</line>
  </lyrics>
  <lyrics ... lang="xxx">
  </lyrics>
</subsonic-response>
```

```json
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.2.0",
    "openSubsonic": true,
    "lyrics": [{
      "artist": "Muse",
      "title": "Hysteria",
      "synced": false,
      "lang": "eng",
      "line": [
        { "value": "It's bugging me" },
        { "value": "Grating me" },
        { "value": "And twisting me around..." }
      ]
    }, {
      "lang": "xxx"
    }]
  }
}
```

##### On the definition of "synced" lyrics
The best source I could find for the LRC format is https://en.wikipedia.org/wiki/LRC_(file_format), which requires the timestamp to be `[mm:ss:xx]` (minutes:seconds: hundredths of a second). However, I would suggest that servers also accept `[mm:ss:xxx]` (minutes:seconds:milliseconds). At the end of the day, the specifics of how a server actually determines this may be out of scope of this extension.

### ID in response
Add an **optional boolean parameter**, `includeIds`. If true, then the response **must** include `artistId` and `id` for the `lyrics` object.

```xml
<subsonic-response status="ok" version="1.2.0" openSubsonic="true">
    <lyrics artist="Muse" artistId="1234" title="Hysteria" id="my-track-id" lang="eng">
        ...
    </lyrics>
</subsonic-response>
```
```json
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.2.0",
    "openSubsonic": true,
    "lyrics": [{
      "artist": "Muse",
      "artistId": "1234",
      "title": "Hysteria",
      "id": "my-track-id",
      "lang": "eng",
      "line": []
    }]
  }
}
```

### All together now
Additional parameters (all optional):
- `id`: string; track id
- ~~`artistId`: string, artist id~~
- ~~`preferSynced`: boolean; this **must** be set to true/false to return the new values. If true, try to return lyrics with timestamp, falling back to regular lyrics~~
- `includeIds`: boolean; if true, add `artistId` and `id` to `lyrics
- `lang`: repeated string; the language(s) to include. If not provided, return all lyrics

Output changes:
- support multiple `lyric`s
- every lyric will have `lang` and `synced` key, may also have `id`/`artistId`
- the body of a lyric is one or more `line`s. If synchronous, each `line` must have a `start` in ms 

### Security impacts

None

### Potential issues

How do the new query parameters interact? Should the backend search just be an AND of all the provided queries? Or, should `artistId` take precedence over `artist`, and `id` take precedence over `title`?

**Comment by epoupon** [comment]
Not sure about `artistId`, I would say an extra `id` parameter to uniquely specify the track, and that overrides all existing fields, should be enough?

**Reply by kgarner7** [reply]
I definitely agree with the `id` part. For `artistId`, it probably makes sense that `artistId` > `artist`?

**Reply by dweymouth** [reply]
Yeah I also don't know why we need `artistId`, since just the `id` should already uniquely identify the song and artist name.

**Reply by kgarner7** [reply]
My initial (wrong) assumption was that an artist/title would be a search. Given that this extension allows returning multiple lyrics, I do think it makes sense to keep artistId if searching is preferred (alternatively, the only query can be by id)

**Reply by dweymouth** [reply]
The only use case I see for artistID is if we want to support returning all known lyrics for a given artist at once. Else, the server can derive the artist name and song name from the song ID, and pass those to Genius or wherever to search for lyrics.

IMO, the client shouldn't care where the lyrics come from, it just says "I want lyrics for this track, prefer synced" and the server either fulfills that from the file tags or online DB as a black box

**Reply by kgarner7** [reply]
That's exactly my thinking; `artist`, `title` and `artistId` are all ways to filter, and `id` is to get the specific track. I guess an alternative would be to have one endpoint `searchLyrics` which takes `artist`/`title`/`artistId`, and `getLyrics2` which only takes `id`.

**Reply by dweymouth** [reply]
Yeah I'd be in favor of the second endpoint for lyrics search then, though I'm still not sure what the client use case would be. (We could always make `getLyrics2` support a list of track IDs to retrieve multiple lyrics in one API call) And does it kind of imply that OpenSubsonic servers must integrate with a lyrics provider (which not every server will want to do), or would it be fine for searchLyrics to return only matches from the local metadata? (.lrc files and ID3 tags)

**Reply by kgarner7** [reply]
I don't see why it implies that the server needs a lyric provider? All of this can just be done with files/tags (as an example, Navidrome just reads the `SYLT`/`USLT` tag and stores that in the database).

I do like the idea of accepting multiple ids for the retrieval.
Honestly, I'm also fine just removing search (`artist`/`title`/`artistId`) entirely. /shrug/

**Reply by Tolriq** [reply]
There's no need for preferSynced since all lyric are returned the client can filter or not use the time values.
And IMO we should only support this for a single song Id. (There's also local .lrc files to support).

The server can have lyrics provider and do tons of things on the UI to gather additional lyrics. But in the end the client just want to display the lyrics for a given song.
There's no real other usage for this endpoint.

**Reply by dweymouth** [reply]
I am good with removing `preferSynced` (since as Tolriq points out synced and unsynced are compatible with the same format), and also supporting just a single track ID. That's good enough for my use case. I guess the only argument for taking multiple trackIDs is if a client wanted to pre-load lyrics for the whole play queue or sections of the play queue, to reduce HTTP round trips

**Reply by Tolriq** [reply]
Not sure loading all lyrics in advance makes sense specially at the start of a queue as it would slow down a lot the display of the current one and use network and server resource slowing down the cache.

Loading next lyric near the end of the track is more efficient.

**Reply by kgarner7** [reply]
I'm fine dropping prefer sync and artist/artistid/title

**Reply by dweymouth** [reply]
What about single vs multiple trackID support?

**Reply by kgarner7** [reply]
I don't have an opinion either way. It would certainly be easier to make this just a single fetch

**Comment by Tolriq** [comment]
If we change the result format then it's a breaking change and we should create a new endpoint.

In that case we should support multiple lyrics and lyrics languages.

**Reply by kgarner7** [reply]
That's fair; let me update with additional information for language support then

**Comment by dweymouth** [comment]
Updated the proposal to reflect the above comments

**Reply by Tolriq** [reply]
Any idea for the name? I think keeping the current X X2 X3 X4 scheme maybe improved for clarity? Since it's for a song getSongLyrics ?

**Reply by dweymouth** [reply]
If we want to move away from the X2,3,4 scheme I think either `getLyricsById` or `getStructuredLyrics` would work as both more clearly communicate the difference between the old and new API in my opinion.

**Reply by Tolriq** [reply]
getLyricsById would imply we have some kind of lyrics ID from the name. Either BySongId or the last one then.

**Reply by dweymouth** [reply]
I'm OK with either of those suggestions. @kgarner7 do you have an opinion here?

**Reply by kgarner7** [reply]
`getLyricsBySongId`? Maybe a bit verbose but it explains exactly what we're getting, and how (for even more verbosity, `getStructuredLyricsBySongId`)

**Reply by dweymouth** [reply]
alright let's go with `getLyricsBySongId` then as the second one seems just a bit too verbose :) Once this is actually launched we can add a note to the other `getLyrics` endpoint saying that most clients should use `getLyricsBySongId` if supported by the server

**Reply by Tolriq** [reply]
Ok and songLyrics for the extension name. 

Now just need to write the PR and get servers onboard.

**Reply by dweymouth** [reply]
ah I guess we need to update the description that the `id` parameter is required (since we got rid of artistID calling this API without `id` doesn't really do anything)

**Comment by duckcantcode** [comment]
Been looking forward to timed lyrics for a while. Super excited about this request. Subscribing.

**Comment by lachlan-00** [comment]
Just to dig up this old discussion, i'm about to start my rewrite of subsonic in Ampache and one thing i'm not going to probably ever include is timed lyrics /LRC /start times of lines.

it's not totally clear here but can i just dump the lyrics in a single [line](https://opensubsonic.netlify.app/docs/responses/line/)?

The lyrics in the database for us is just a single varchar list of text with newlines and i'd prefer to keep it like that where possible

**Reply by Tolriq** [reply]
No you need to split on new lines and return the array

```
"line": [
        { "value": "It's bugging me" },
        { "value": "Grating me" },
        { "value": "And twisting me around..." }
      ]
```
This is important for better display client side and empty lines handling.

**Reply by lachlan-00** [reply]
k, good to know i can handle that easily
